### PR TITLE
Avoid setting model.df_model in results.__init__

### LINKED ISSUE #3734: BUG: mutable model.df_model in discrete_model
Motivated by #3727:

> The k_extra attribute is for some generic, inherited methods, I don't remember which.

`k_extra` modifies `FooResults.df_model` and `df_resid`, but handling `k_extra` in a consistent manner requires a bit of kludge-work.  In the process of writing out _why_ it requires kludge-work, I found that a pet peeve was actually a bug.

Each of the `L1FooResults` classes alter `model.df_model` and `df_resid` within their `__init__` methods.  I think this should be avoided, as `model.df_foo` should be essentially immutable.  The problem caused by mutability is this:

```
>>> data = sm.datasets.randhie.load()
>>> model = sm.NegativeBinomial(data.endog, data.exog, loglike_method='nb2')
>>> res1 = model.fit()
>>> res2 = model.fit_regularized(alpha=100)
>>> res3 = model.fit()
>>> res1.df_model
8.0
>>> res3.df_model
6
>>> (res1.aic, res3.aic)
(87507.710855394151, 87503.710855394151)
```

`res1` and `res3` should be identical, but are not because the L1FooResults instance is modifying model.df_bar.  

In my local "screw backward compatibility" branch, the closest thing I've found to a solution is to define the Results-level `k_extra`, `df_model`, and `df_resid` as properties/cache_readonlys that get mixed in to DiscreteResults:

```
class DiscreteResultsFreedomMixin(object):

	@cache_readonly
	def k_extra(self):
		model = self.model
		k_extra = model.k_extra
		if getattr(model, 'loglike_method', '').startswith('nb'): ## FIXME: kludge
			k_extra -= 1
		return k_extra

	@cache_readonly
	def df_model(self):
		k_constr = self.k_constr
		k_extra = self.k_extra

		ret = self.model.df_model - (k_constr + k_extra)
		return ret

	@cache_readonly
	def df_resid(self):
		k_constr = self.k_constr
		k_extra = self.k_extra

		ret = self.model.df_resid + (k_constr + k_extra)
		return ret
```

This avoids the need to alter model.df_foo (which gets set only in `DiscreteModel.initialize` and `MultinomialModel.initialize`), but I'm not wild about the kludge that sets `k_extra`, in part because I'm not sure it is generally-correct as opposed to doesn't-cause-test-failures-correct.

### PR DESCRIPTION
**Pull request by jbrockmendel**
Test case to check that model.df_model is not changed on call to fit_regularized

closes #3734

### CODE REVIEW

File: statsmodels/discrete/tests/test_discrete.py [hunk 1]
```diff
@@ -1559,6 +1559,28 @@ def test_binary_pred_table_zeros():
     assert_equal(res.pred_table(), expected)
 
 
+def test_unchanging_degrees_of_freedom():
+    # see GH3734
+    data = sm.datasets.randhie.load()
+    model = sm.NegativeBinomial(data.endog, data.exog, loglike_method='nb2')
+    
+    res1 = model.fit()
```
Comment (josef-pkt) [inline statsmodels/discrete/tests/test_discrete.py]: in general: If convergence is not of interest in a test, then use start_params with solution, or close to solution.
that saves some time in the test runs

File: statsmodels/discrete/tests/test_discrete.py [hunk 2]
```diff
@@ -1559,6 +1559,28 @@ def test_binary_pred_table_zeros():
...
+    res2 = model.fit_regularized(alpha=100)
+    assert_(res2.df_model != 8)
+    # The case of interest is that when res2.df_model != res1.df_model.
+    # Make sure this is the case before making the assertion we actually
+    # care about.
+
+    res3 = model.fit()
+    # res3 should be identical to res1.  This test is being added in a PR
+    # that fixes a bug in which the call to `fit_regularized` modifies
+    # model.df_model, leading to res3.df_model != res1.df_model
```
Comment (josef-pkt) [inline statsmodels/discrete/tests/test_discrete.py]: this is not supposed to be a novel :)
one line (or maybe more for essential information) should be enough, details are in the issue.
e.g. "df should not be changed by having called fit_regularized"

also you can also check the model df directly, e.g.
`assert_equal(res2.model.df_model, res1.model.df_model)`

Comment (jbrockmendel) [inline statsmodels/discrete/tests/test_discrete.py]: A lot of authors weren't appreciated in their own eras...

> also you can also check the model df directly, e.g.
assert_equal(res2.model.df_model, res1.model.df_model)

Aren't `res2.model` and `res1.model` the same object?

Comment (josef-pkt) [inline statsmodels/discrete/tests/test_discrete.py]: yes, you're right.
the res2.model.df_model check needs to assert the integer, or a copy of the df

**Comment by jbrockmendel** [comment]
Related: #1203

**Comment by jbrockmendel** [comment]
Looks like similar incrementing of df_model and df_resid happens in generalized_linear_model and in `base._constraints.fit_constrained_wrap`.  

The same fix here can be applied to genmod.  Unfortunately that method is not getting covered when I run tests locally.  Coveralls says it is covered; I'll look into the discrepancy.

The version in _constraints is called only from tests, and is a near copy/paste of the methods in discrete and genmod, so I'm guessing it is on its way out.

**Comment by josef-pkt** [comment]
Are you running the full test suite?
The genmod test are a bit spread out across subfolders, depending a bit on where I added them during an extension, e.g. there are several in discrete and in base for the semi-generic functionality, e.g. cov_type tests are in several places and not always where the corresponding model is, e.g. writing one set of test for both Poisson and GLM-Poisson or OLS and GLM-Gaussian.

The version in statsmodels.base are the most generic solutions, e.g. IIRC Poisson -> GLM -> generic was the development path. The main problem for adding base._constrained fit_constrained to all models is that it needs offset, which was or is only available in GLM and discrete.CountModels. Enhancement to other models are still missing. So. base._constraints is waiting for the subclasses to catch up before taking over.
(typical development for generic features: try it out for one model, extend it to similar models, try to figure out what is required by models for this to work, make up some API along the way, and finally add the generic solution, if we manage to figure it out, to a base class.)

File: statsmodels/discrete/tests/test_discrete.py (hunk 2 above)
Comment (jbrockmendel) [inline statsmodels/discrete/tests/test_discrete.py]: I'm not entirely clear: does your last comment mean there is still an edit needed?

**Comment by jbrockmendel** [comment]
>  Are you running the full test suite?

I must not have been.  Now it is getting hit as expected.

> So. base._constraints is waiting for the subclasses to catch up before taking over.

Luckily the code there doesn't alter self.model.df_foo inplace, only increments self._results.df_foo.  In a perfect world I'd like to avoid even doing that much, but I'm not going to sweat this peripheral case for now.

File: statsmodels/discrete/tests/test_discrete.py (hunk 2 above)
Comment (josef-pkt) [inline statsmodels/discrete/tests/test_discrete.py]: You can remove the comment, but I was thinking that a direct test the model.df_model remains == 8 and similar for df_resid would be the more direct test that it is unchanged.

Comment (jbrockmendel) [inline statsmodels/discrete/tests/test_discrete.py]: I guess so.  The assertion regarding `res2.df_model` is needed to ensure we don't accidentally have a trivially-passing test.  We could even skip the `fit()` calls if we really wanted.  I'm inclined to keep it as is because of a very specific laziness.

**Comment by josef-pkt** [comment]
> Luckily the code there doesn't alter self.model.df_foo inplace

Luck or design?  In some parts we can still see the age of the code. In the early development of statsmodels it wasn't so clear that we have to be careful about whether attributes are immutable model attributes or mutable results attributes. discrete fit_regularized hasn't been refactored since the beginning quite some time ago. (And that's another reason that a code review every 5 (?) years is useful.)

In a near perfect world, I can still argue that df_xxx are attributes of an instance and are supposed to be set in `__init__` (of results).

--- 

another special/generic case
https://github.com/statsmodels/statsmodels/blob/master/statsmodels/base/model.py#L2057
where I retrofit model.df_xxx if it doesn't exist.
This is supposed to handle the case where we only know in fit and after how many parameters we actually have. It also will have cases that it will not handle correctly, but it is difficult to figure out how many parameters we have in general nonlinear models .

**Comment by jbrockmendel** [comment]
> where I retrofit model.df_xxx if it doesn't exist

Yah, I stumbled on that a few minutes ago.  Looks like the only place it gets hit is in `tsa.arma_mle`.  That has the long-standing problem of having `k_ar` and `k_ma` (well actually `nar` and `nma`) not being set at `__init__`, so it doesn't make sense to set its `df_model` and `df_resid` attributes.

I propose we patch arma_mle so that it doesn't reach that branch, then require that model.df_model be defined before the call to `fit`, and get rid of the not-hasattr branches there  (possibly after a release cycle in which those branches issue a DeprecationWarning or something).

> In a near perfect world, I can still argue that df_xxx are attributes of an instance and are supposed to be set in __init__ (of results).

The more I look at this, the more I think that is feasible.  Moreover, the shim that will make it work may also take care of the "#create dummy results Instance, TODO: wire up properly" [workaround](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/discrete/discrete_model.py#L1069) in all three appearances of `fit_constrained`.

The only cases I see where `results.df_model != model.df_model` are either a) `fit_regularized`, b) `fit_constrained`.

The L1 case is pretty easy, let's focus on the constrained case.  We can pass `k_constr` (and possibly `res_constr`) as kwargs to `fit` inside `fit_constrained`.  A two line patch to `LikelihoodModel.fit` will pass that through to `LikelihoodModelResults.__init__`.  Then in `__init__` set `results.df_model = model.df_model - kwargs.get('k_constr', 0)`.  Similar for `df_resid`. 

I'll try it and check that this actually works.

**Comment by josef-pkt** [comment]
GenericLikelihoodModel and associated Results are allowed and supposed to have some "magic" so it works for different usecases without requiring the user writing a MLE Model as subclass to figure out and set many details.
miscmodels.TModel and TArma are examples. TModel allows the df of the t-distribution to be fixed or estimated without much extra coding effort.

I don't really like hardcoding and imposing something like k_constr in the base classes. Any assumptions and pattern that are hard coded based on current models might make future extensions more difficult
https://github.com/statsmodels/statsmodels/pull/1665/files#diff-204b0120deaf8dc3d3037809c40cc9b1L154

**Comment by jbrockmendel** [comment]
> I don't really like hardcoding and imposing something like k_constr in the base classes. Any assumptions and pattern that are hard coded based on current models might make future extensions more difficult

Suppose hypothetically that the suggestion 1) does work for all existing cases and 2) reflects an underlying logic that tentatively appears to apply generally.  What would an appropriate place for that be?

**Comment by josef-pkt** [comment]
I think a relatively simple way will be to have eventually a proper `results.__init__`, e.g. df_resid and df_model could be added as keyword arguments, and fit is responsible to construct them based on the fit information and keywords. (One trade-off is that df_resid is not necessarily needed, e.g. for predict, and it should be lazily evaluated if it is costly to compute, as in the case of the trace of the hat-matrix in Ridge or other penalized regression.)

A related problem with a large amount of retrofitting the results instance is
https://github.com/josef-pkt/statsmodels/commit/d1c578391169fe896527bd41ce3b67ea991aa0ec#diff-b165b4bd4e10edd0dfb485e47c562b2cR581
issue #2393

The problem is that the results classes were initially designed to be closely linked to a specific model class. Essentially, it is one class split into two to take advantage of inheritance (and to keep results attributes separate from model attributes).
As we accumulate more variation in the models and estimation methods this tight integration becomes a serious limitation, unless we keep adding new results classes for every variation. (e.g. using Mixins more heavily to combine different components.)

also related to https://github.com/statsmodels/statsmodels/issues/3738#issuecomment-305988137

**Comment by jbrockmendel** [comment]
Between this thread and #3738, it seems like there are two inter-related topics here:

1) How should XYZResults be created, inherited, wrapped, and have their attributes set.

2) How should `df_model` and `df_resid` be defined/set.

(And at the intersection of these two, the actual topic of the PR, in-place modification of model attributes outside model methods.)

On the first, I'm resigned to punting on creation/inheritance/wrapping as being not worth the effort to try to push significant design changes.  A better use of time will be incrementally moving attribute setting into `__init__` (or `initialize` in some cases) in cases where doing so leaves the logic unchanged.

On the second, in a perfect world df_model and df_resid would be functions of model attributes, data/params shapes etc in a way that was internally consistent-- i.e. a natural fit for properties.  I'm giving up on the properties suggestion for now in favor of more incremental adjustments towards systematization.

As to the PR itself, I think the only outstanding question is whether the same bugfix should be applied to GenericLikelihoodModelResults.  I'd venture "yes", but would settle for a "# TODO: consider changing this" note and kicking it down the road.

**Comment by josef-pkt** [comment]
GenericLikelihoodModelResults

```
        if hasattr(model, 'df_resid'):
            self.df_resid = model.df_resid
        else:
            self.df_resid = self.endog.shape[0] - self.df_model
            # retrofitting the model, used in t_test TODO: check design
            self.model.df_resid = self.df_resid
```

This does not modify an existing model.df_resid, so it should be fine.
However, this will or would not be necessary if the methods in results don't access results.model.df_resid directly anymore.  I don't know if all models/results classes have now df_resid as a results attribute or property.

I had forgotten that this discussion is in a PR, given the simultaneous discussion in issues.
The changes in this PR are good and just a non-controversial bug fix.

Any more general changes will need to check and make sure that df_resid, df_model have enough test coverage across models. df_resid is not used in the basic results with asymptotic normal assumption, i.e. when use_t=False, So the basic unit tests for a model will or might not always cover it.

`t_test` enforces that df_resid is defined, but doesn't use it's value if use_t=False.

**Comment by jbrockmendel** [comment]
> This does not modify an existing model.df_resid, so it should be fine.

You're right, but it is modifying the `model` object.  Changing what attributes the model *has* may not cause the same types of problems as re-setting them, but it definitely makes development more difficult.  Having to track down and/or guess what attributes exist is sub-optimal.

> However, this will or would not be necessary if the methods in results don't access results.model.df_resid directly anymore. I don't know if all models/results classes have now df_resid as a results attribute or property.

In the tests, the only model that reaches that code is `tsa.arma_mle`.  In some ways arma_mle would not be out of place in miscmodels.  It doesn't need df_model or df_resid internally.

> Any more general changes will need to check and make sure that df_resid, df_model have enough test coverage across models. df_resid is not used in the basic results with asymptotic normal assumption, i.e. when use_t=False, So the basic unit tests for a model will or might not always cover it.

Let's wrap this up and move this part of the discussion to the next PR/Issue.

**Comment by josef-pkt** [comment]
> just a quick check, hitting a line of code in the tests doesn't mean that the attribute is necessary.

base t_test uses
`df_resid = getattr(self, 'df_resid_inference', self.df_resid)`

I don't find any search result for `self.model.df` in base.model except for the setting of the df in the model instance.
This would indicate that `model.df_resid` is not accessed in the base models and therefore shouldn't be needed in GenericLikelihoodModels. (sandwich_covariance for cov_type also doesn't use any model.df_xxx.)
So, we are already pretty far in using only the results.df, and the model.df are only needed for setting the results.df.

**Comment by jbrockmendel** [comment]
@josef-pkt I think we've reached an agreement on the actual content of this PR.  Some of the discussion above is on related topics and can be moved to an Issue.

I'll rebase and de-conflict once you give the go-ahead.

**Comment by josef-pkt** [comment]
looks good, 

needs rebase to remove "Merge branch 'master' ..."
I think we can do a full squash in rebase. 
merge is on my todo list for tomorrow

**Comment by jbrockmendel** [comment]
Excellent!  Really psyched to see this one go through
